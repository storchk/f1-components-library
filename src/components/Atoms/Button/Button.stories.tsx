import type { Meta, StoryObj } from '@storybook/react'

import { Button } from './Button'
import type { ButtonProps } from './Button.types'

export default {
  title: 'Atoms/Button',
  component: Button,
  args: {
    color: 'primary',
    label: 'Button',
  },
  argTypes: {
    color: {
      options: ['black', 'gold', 'white', 'primary'],
      control: {
        type: 'select',
      },
    },
  },
} as Meta

type Story = StoryObj<ButtonProps>

export const Default: Story = {
  render: props => <Button variant="default" {...props} />,
}
export const Positive: Story = {
  render: props => <Button variant="positive" {...props} />,
}
export const TextOnly: Story = {
  render: props => <Button variant="textOnly" {...props} />,
}
export const TextLink: Story = {
  render: props => <Button variant="textOnly" {...props} />,
}
