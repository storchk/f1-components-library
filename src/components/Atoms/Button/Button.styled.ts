import styled, { css } from 'styled-components'

import type { ThemeType } from '../../../theme'
import type { ButtonColor, StyledButtonProps } from './Button.types'

const getColorStylesByColor = (theme: ThemeType, color: ButtonColor) => {
  switch (color) {
    case 'black':
      return css`
        border-color: ${theme.colors.black};
        background-color: ${theme.colors.black};

        &:hover {
          background-color: ${theme.colors.textDarkSoft};
          border-color: ${theme.colors.white};
        }
      `
    case 'gold':
      return css`
        border-color: ${theme.colors.gold};
        background-color: ${theme.colors.gold};

        &:hover {
          background-color: ${theme.colors.goldLight};
          border-color: ${theme.colors.goldLight};
        }
      `
    case 'white':
      return css`
        border-color: ${theme.colors.white};
        background-color: ${theme.colors.white};

        &:hover {
          background-color: ${theme.colors.ceramic};
          border-color: ${theme.colors.ceramic};
        }
      `
    case 'primary':
    default:
      return css`
        border-color: ${theme.colors.primary};
        background-color: ${theme.colors.primary};

        &:hover {
          background-color: ${theme.colors.primaryLight};
          border-color: ${theme.colors.primaryLight};
        }
      `
  }
}

export const StyledButton = styled.button<StyledButtonProps>(
  ({ theme, $variant, $color }) => css`
    padding: ${theme.spacing.xs} ${theme.spacing.sm};
    border-radius: ${theme.border.radius};
    border: ${theme.border.width} solid;
    color: ${$color === 'gold' ? theme.colors.textDark : theme.colors.textWhite};

    ${$variant === 'positive' ? css`` : ''}
    ${getColorStylesByColor(theme, $color)}
  `
)
