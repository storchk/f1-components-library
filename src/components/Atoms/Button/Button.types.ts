import type { AnchorHTMLAttributes } from 'react'

import type { ThemeType } from '../../../theme'

export type ButtonColor = 'primary' | 'secondary'
export type ButtonVariant = 'primary' | 'secondary' | 'ghost' | 'text'

export type ButtonProps = {
  color?: ButtonColor
  variant?: ButtonVariant
  isLoading?: boolean
  label: string
} & AnchorHTMLAttributes<HTMLAnchorElement | HTMLButtonElement>

export type StyledButtonProps = {
  $color: ButtonColor
  $variant: ButtonVariant
}

export type GetButtonVariantsProps = {
  theme: ThemeType
} & StyledButtonProps
