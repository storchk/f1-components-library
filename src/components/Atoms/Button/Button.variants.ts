import { css } from 'styled-components'

import type { GetButtonVariantsProps } from './Button.types'

export const getButtonVariants = ({ theme, $color, $variant }: GetButtonVariantsProps) => {
  switch ($variant) {
    case 'primary':
    default:
      return css`
        color: ${theme.colors.secondary};
        border-color: ${theme.colors[$color]};

        &:hover {
          box-shadow: 0 0 0 2px ${theme.colors[$color]} inset;
        }
      `
    case 'secondary':
      return css`
        color: ${theme.colors.pureWhite};
        border-color: ${theme.colors[$color]};
        background-color: ${theme.colors[$color]};

        ${$color === 'primary'
          ? css`
              &:hover {
                color: ${theme.colors.secondary};
                border-color: ${theme.colors[$color]};
                background-color: transparent;
              }
            `
          : css`
              &:hover {
                border-color: ${theme.colors[`${$color}Accent`]};
                background-color: ${theme.colors[`${$color}Accent`]};
              }
            `}
      `
    case 'ghost':
      return css`
        color: ${theme.colors[$color]};
        border: none;

        &:hover {
          color: ${theme.colors[`${$color}Accent`]};
        }
      `
    case 'text':
      return css`
        color: ${theme.colors[$color]};
        padding: 0;
        border: none;
        border-radius: 0;

        &:hover {
          color: ${theme.colors[`${$color}Accent`]};
        }
      `
  }
}
