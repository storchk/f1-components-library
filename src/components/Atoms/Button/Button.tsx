import { Typography } from '../Typography'
import { StyledButton } from './Button.styled'
import type { ButtonProps } from './Button.types'
export const Button = ({
  variant = 'primary',
  color = 'primary',
  label,
  onClick,
}: ButtonProps): JSX.Element => {
  return (
    <StyledButton $variant={variant} $color={color} onClick={onClick}>
      <Typography $isUppercase fontSize="xs" fontWeight="semibold">
        {label}
      </Typography>
    </StyledButton>
  )
}
