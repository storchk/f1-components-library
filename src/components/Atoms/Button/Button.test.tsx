import { testing } from '../../../testing'
import { Button } from './Button'
import { ButtonProps } from './Button.types'
import { axe } from 'jest-axe'

const { render, screen, renderer } = testing
const label = 'Hello world! I am using React'

const renderComponent = (props?: Partial<ButtonProps>) =>
  render(<Button label={label} {...props} />)
describe('Atoms: Button', () => {
  it('should render button with text', () => {
    renderComponent()
    const text = screen.getByText(/Hello world! I am using React/i)
    expect(text).toBeInTheDocument()
  })

  it('snapshot | variant: primary', () => {
    const view = renderer(<Button variant="primary" label={label} />)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | variant: secondary', () => {
    const view = renderer(<Button variant="secondary" label={label} />)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | variant: ghost', () => {
    const view = renderer(<Button variant="ghost" label={label} />)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | variant: test', () => {
    const view = renderer(<Button variant="text" label={label} />)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | color: primary', () => {
    const view = renderer(<Button color="primary" label={label} />)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | color: secondary', () => {
    const view = renderer(<Button color="secondary" label={label} />)
    expect(view).toMatchSnapshot()
  })

  test('A11y - No violations detected', async () => {
    const { container } = renderComponent()
    expect(await axe(container)).toHaveNoViolations()
  })
})
