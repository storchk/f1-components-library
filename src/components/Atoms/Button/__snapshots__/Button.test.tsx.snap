// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms: Button snapshot | color: primary 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #1f1f27;
  border-color: #e10600;
}

.c0:hover {
  box-shadow: 0 0 0 2px #e10600 inset;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;

exports[`Atoms: Button snapshot | color: secondary 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #1f1f27;
  border-color: #1f1f27;
}

.c0:hover {
  box-shadow: 0 0 0 2px #1f1f27 inset;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;

exports[`Atoms: Button snapshot | variant: ghost 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #e10600;
  border: none;
}

.c0:hover {
  color: #c30500;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;

exports[`Atoms: Button snapshot | variant: primary 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #1f1f27;
  border-color: #e10600;
}

.c0:hover {
  box-shadow: 0 0 0 2px #e10600 inset;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;

exports[`Atoms: Button snapshot | variant: secondary 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #fff;
  border-color: #e10600;
  background-color: #e10600;
}

.c0:hover {
  color: #1f1f27;
  border-color: #e10600;
  background-color: transparent;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;

exports[`Atoms: Button snapshot | variant: test 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 600;
  color: inherit;
  font-family: Titillium Web;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 24px;
  border-radius: 4px;
  border: 2px solid;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  color: #e10600;
  padding: 0;
  border: none;
  border-radius: 0;
}

.c0:hover {
  color: #c30500;
}

<button
  className="c0"
>
  <span
    className="c1"
    fontSize="xs"
    fontWeight="semibold"
  >
    Hello world! I am using React
  </span>
</button>
`;
